### ISSUE TITLE
v2 compatible on full Kubeflow

### ISSUE THREAD

**Issue by Bobgy** [kubeflow/pipelines#5680]
UPDATE:
PRs have been written:
* https://github.com/kubeflow/pipelines/pull/5697
* https://github.com/kubeflow/pipelines/pull/5750
* https://github.com/kubeflow/gcp-blueprints/pull/284

===

This issue tracks efforts to make [v2 compatible pipelines](https://www.kubeflow.org/docs/components/pipelines/sdk/v2/) run on [full Kubeflow](https://www.kubeflow.org/docs/components/pipelines/installation/overview/#full-kubeflow-deployment).

Depends on #4649

The major differences from KFP standalone are:
* MLMD grpc service is in kubeflow namespace by default.
* MinIO service is in kubeflow namespace by default.
* In Kubeflow on GCP, the default bucket is on GCS, so it's not `mlpipeline`.

For MLMD grpc service, we are already reading a configmap called `metadata-grpc-configmap`. The configmap is already configured in full Kubeflow for each user namespace: https://github.com/kubeflow/pipelines/blob/7608fcdbb4210c153b4a97b019787c6291473b9f/manifests/kustomize/base/installs/multi-user/pipelines-profile-controller/sync.py#L52-L63
Therefore, it already works out of the box.

Because in Kubeflow on GCP, the default bucket is not `mlpipeline`, we need configurations that can set default pipeline root and credentials for each namespace, this is essentially https://github.com/kubeflow/pipelines/issues/4649 for v2 compatible pipelines.

The following comments are investigation logs towards the conclusion I summarized here.

**Comment by Bobgy** [comment]
I'm trying to run v2 compatible pipeline on full Kubeflow, as expected, it fails when uploading to minio storage:
```
F0519 07:16:43.219511       1 main.go:56] Failed to execute component: failed to upload output artifact "output_dataset_one" to remote storage URI "minio://mlpipeline/v2/artifacts/two_step_pipeline/two-step-pipeline-z9pxp/preprocess/output_dataset_one": uploadFile(): unable to complete copying "/minio/mlpipeline/v2/artifacts/two_step_pipeline/two-step-pipeline-z9pxp/preprocess/output_dataset_one" to remote storage "two_step_pipeline/two-step-pipeline-z9pxp/preprocess/output_dataset_one": failed to close Writer for bucket: blob (key "two_step_pipeline/two-step-pipeline-z9pxp/preprocess/output_dataset_one") (code=Unknown): RequestError: send request failed
caused by: Put "
http://minio-service:9000/mlpipeline/two_step_pipeline/two-step-pipeline-z9pxp/preprocess/output_dataset_one
": dial tcp: lookup minio-service on 10.11.240.10:53: no such host
```

**Comment by Bobgy** [comment]
Interesting, when I switch pipeline_root to a gcs path like gs://gongyuan-dev/v2/.
The pipeline runs successfully, so finding mlmd grpc service host was not a problem.

I checked how that's configured -- the mlmd grpc service host and port come from k8s env vars:

```
command:
...
- '--mlmd_server_address'
- $(METADATA_GRPC_SERVICE_HOST)
- '--mlmd_server_port'
- $(METADATA_GRPC_SERVICE_PORT)
```

**Comment by Bobgy** [comment]
it seems that the easiest way to make minio service host/port auto discovery work is to switch to the same mechanism

**Comment by Bobgy** [comment]
Because configmaps needs to be deployed, I propose a simpler default mechanism:
1. Find if env vars like `MINIO_SERVICE_SERVICE_HOST` and `MINIO_SERVICE_SERVICE_PORT` exists, if they do, then we are running in KFP single namespace mode, we can use these env vars to connect to minio service.
2. If the env vars do not exist, we can assume we are running in KFP multi user mode, so default minio service should be `minio-service.kubeflow:9000`.
3. If that fails, we'll just tell users it fails

Going forward, a proper configuration for artifact storage will need to be introduced -- that's when we can make this configurable.

**Comment by Bobgy** [comment]
So actually, configurability of credentials is not a must have for making v2compat run on full Kubeflow.

It's enough to make default pipeline root configurable. For KF on GCP, we can default to gcs path. For others, using minio://mlpipeline is enough.

**Comment by Bobgy** [comment]
/reopen
we need to keep track of integration with Kubeflow 1.4

**Comment by Bobgy** [comment]
I think one issue that I previously missed is that, we need to make the new task API available from pipeline pods. That requires some istio authorization policy change.

**Comment by Bobgy** [comment]
TODOs:
* [ ] expose task API via Istio authorization policy
* [ ] merge https://github.com/kubeflow/gcp-blueprints/pull/284
